libraries/botframework-schema/src/activityEx.ts

```diff
@@ -0,0 +1,405 @@
+   */
+    export function hasContent(source: Partial<Activity>): boolean  {
+        if(source.text !== undefined && source.text.trim().length > 0) {
+            return true;
+        }
+
+        if(source.summary !== undefined && source.summary.trim().length > 0) {
+            return true;
+        }
+
+        if(source.attachments !== undefined && source.attachments.length > 0) {
+            return true;
+        }
+
+        if(source.channelData !== undefined) {
+            return true;
+        }
+
+        return false;
+    }
+
+  /**
+   * Resolves the mentions from the entities of the source activity.
+   * @param source The source activity.
+   * @returns The array of mentions; or an empty array, if none are found.
+   * @remarks This method is only intended for use with a message activity, where the Activity Type is set to Message.
+   * @see cref="entities" .
+   * @see cref="mention" .
+   */
+    export function getMentions(source: Partial<Activity>) : Mention[] {
+        return source.entities.filter(x => x.type.toLowerCase() === 'mention')
+        .map(e => e as Mention);
+    }
+
+  /**
+   * Creates a ConversationReference based on the source activity.
+   * @param source The source activity.
+   * @returns A conversation reference for the conversation that contains the activity.
+   */
+    export function getConversationReference(source: Partial<Activity>) : Partial<ConversationReference> {
+        return {
+            activityId: source.id,
+            user: source.from,
+            bot: source.recipient,
+            conversation: source.conversation,
+            channelId: source.channelId,
+            locale: source.locale,
+            serviceUrl: source.serviceUrl
+        } as ConversationReference;
+    }
+
+  /**
+   * Determines if the Activity was sent via an Http/Https connection or Streaming.
+   * This can be determined by looking at the ServiceUrl property:
+   * (1) All channels that send messages via http/https are not streaming
+   * (2) Channels that send messages via streaming have a ServiceUrl that does not begin with http/https.
+   * @param source The source activity.
+   * @returns True if the Activity was originate from a streaming connection.
+   */
+    export function isFromStreamingConnection(source: Partial<Activity>) : boolean {
+        if (source.serviceUrl !== undefined) {
+            const isHttp: boolean = source.serviceUrl.toLowerCase().startsWith("http");
+            return !isHttp;
+        }
+        return false;
+    }
+
+  /**
+   * Indicates whether this activity is of a specified activity type.
+   * @param source The source activity.
+   * @param activityType The activity type to check for.
+   * @returns True if the activity is of the specified activity type; otherwise, false.
+   */
+    export function isActivity(source:Partial<Activity>, activityType: string): boolean {

```

Comment:
Where is the test coverage for this function? What happens when `typeof source.type !== 'string'`?

Reply:
Hi @stevengum, this code was covered by the tests of the `asXXXActivity` methods as it is called by them. However, we added new tests to cover all possibilities.

Reply:
Sorry, that of course tracks. However, thank you for adding individual tests as the `asXXXActivity` and `isActivity` methods are both public.